transport/x/grpc/inbound.go

```diff
@@ -0,0 +1,178 @@
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.router = router
+}
+
+// Transports implements transport.Inbound#Transports.
+func (i *Inbound) Transports() []transport.Transport {
+	return []transport.Transport{}
+}
+
+func (i *Inbound) start() error {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	if i.router == nil {
+		return errRouterNotSet
+	}
+	serviceDescs, err := i.getServiceDescs()
+	if err != nil {
+		return err
+	}
+	server := grpc.NewServer(
+		grpc.CustomCodec(customCodec{}),
+		// TODO: does this actually work for yarpc
+		// this needs a lot of review
+		grpc.UnaryInterceptor(otgrpc.OpenTracingServerInterceptor(i.inboundOptions.getTracer())),
+	)
+	for _, serviceDesc := range serviceDescs {
+		server.RegisterService(serviceDesc, noopGrpcStruct{})
+	}
+	go func() {
+		// TODO there should be some mechanism to block here
+		// there is a race because the listener gets set in the grpc
+		// Server implementation and we should be able to block
+		// until Serve initialization is done
+		//
+		// It would be even better if we could do this outside the
+		// lock in i
+		//
+		// TODO Server always returns a non-nil error but should
+		// we do something with some or all errors?
+		_ = server.Serve(i.listener)
+	}()
+	i.server = server
+	return nil
+}
+
+func (i *Inbound) stop() error {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	if i.server != nil {
+		i.server.GracefulStop()
+	}
+	return nil
+}
+
+func (i *Inbound) getServiceDescs() ([]*grpc.ServiceDesc, error) {
+	// TODO: router.Procedures() is not guaranteed to be immutable
+	procedures := i.router.Procedures()
+	if len(procedures) == 0 {
+		return nil, errRouterHasNoProcedures
+	}
+	serviceNameToServiceDesc := make(map[string]*grpc.ServiceDesc)
+	for _, procedure := range procedures {
+		serviceName, methodDesc, err := i.getServiceNameAndMethodDesc(procedure)
+		if err != nil {
+			return nil, err
+		}
+		serviceDesc, ok := serviceNameToServiceDesc[serviceName]
+		if !ok {
+			serviceDesc = &grpc.ServiceDesc{
+				ServiceName: serviceName,
+				HandlerType: (*noopGrpcInterface)(nil),
+			}
+			serviceNameToServiceDesc[serviceName] = serviceDesc
+		}
+		serviceDesc.Methods = append(serviceDesc.Methods, methodDesc)
+	}
+	serviceDescs := make([]*grpc.ServiceDesc, 0, len(serviceNameToServiceDesc))
+	for _, serviceDesc := range serviceNameToServiceDesc {
+		serviceDescs = append(serviceDescs, serviceDesc)
+	}
+	return serviceDescs, nil
+}
+
+func (i *Inbound) getServiceNameAndMethodDesc(procedure transport.Procedure) (string, grpc.MethodDesc, error) {
+	serviceName, methodName, err := procedureNameToServiceNameMethodName(procedure.Name)
+	if err != nil {
+		return "", grpc.MethodDesc{}, err
+	}
+	return serviceName, grpc.MethodDesc{
+		MethodName: methodName,
+		// TODO: what if two procedures have the same serviceName and methodName, but a different service?
+		// TODO: should we handle procedure.Encoding somehow?
+		Handler: newHandler(
+			procedure.Service,
```

Comment:
I think this something we need for vanilla grpc, I need this is general in the past, I think we'll need to address this as we address general yarpc/grpc routing issues.